### PR TITLE
build: add commit msg template with Release note.

### PR DESCRIPTION
**Pull request by jordanlewis**
Also, add a commit-msg git hook that warns if you didn't include a
release note at all.

Release note: None

### CODE REVIEW

**Comment by jordanlewis** [comment]
The downside to this approach is that it'll break things if you switch to a branch that doesn't have this patch and try to commit, since git won't be able to find the template file. Open to suggestions, but that's why the git hook is better.

File: githooks/commit-msg [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+if ! grep '^Release note.*: ' "$1" 2>&1 >/dev/null ; then
```
Comment (benesch) [inline githooks/commit-msg]: `grep -q`?

Comment (benesch) [reply githooks/commit-msg]: I also wouldn't mind being a bit more stringent:

```
grep -qE '^Release note( \([^)]+\))?: '
```

File: Makefile [hunk 2]
```diff
@@ -302,6 +302,10 @@ $(GITHOOKSDIR)/%: githooks/%
 	@ln -s ../../$(basename $<) $(dir $@)
 endif
 
+.PHONY: commit_template
+commit_template:
+	@git config --local commit.template build/commit-template
```
Comment (benesch) [inline Makefile]: Agreed that the commit hook approach is better.

**Comment by tschottdorf** [comment]
I like having the warning. Left a suggestion about the commit template, which we ought to encourage.
Were you going to submit the lint again next?

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[Makefile, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-Xa--zAFPfLMMJ2UAb-r1-307:-L-Y1ow9guJ3gpTpLukr:bnbkgc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd4c98afbae85e29065be186779bbe81fb115b10/Makefile#L307)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Agreed that the commit hook approach is better.
</blockquote></details>

Yeah, hadn't thought about the branch switching. Having a commit template is generally still really nice. It's a shame `git commit` fails spectacularly when the file isn't found.

As a compromise, perhaps adding to the warning in the `commit-msg`code could prove useful:

```bash
tplfile="${GIT_DIR}/.gitmessage"
if [ ! -e "${tplfile}" ]; then
  cat<<EOF
Run the following to install a sane commit message template for this repository:

    build/commit-template ${tplfile} && git config --local commit.template ${tplfile}
EOF   
fi
```

---

*[build/commit-template, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-Y2tDtGnp-C3pFnsSA:-L-Y2tDuwAz6AtmPJNyo:bnb95zm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd4c98afbae85e29065be186779bbe81fb115b10/build/commit-template#L1)):*
> ```
> 
> ```

Might as well add the headline: `<package>: <short description>`

---

*[build/commit-template, line 7 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-Y1lJPcDgnWDdvOvHZ:-L-Y1lJPcDgnWDdvOvH_:bgf3by8) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd4c98afbae85e29065be186779bbe81fb115b10/build/commit-template#L7)):*
> ```
> # changes. Leave the message above if not. For example:
> #
> # Release note (sql change): Add support for JSON columns.
> ```

Mind changing `CONTRIBUTING.md` to link to this file so that we don't duplicate?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-:-L-Y2SImy0dGpnRevgtr:b-xf8bwf)*

**Comment by jordanlewis** [comment]
I'm planning to leave the lint out for a week or two to see how we do without it.

---

Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[githooks/commit-msg, line 3 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-Xa-0-S2egfljYlGI6-r1-3:-L-YP3vmiS18Kxq5gWd-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd4c98afbae85e29065be186779bbe81fb115b10/githooks/commit-msg#L3)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I also wouldn't mind being a bit more stringent:

```
grep -qE '^Release note( \([^)]+\))?: '
```
</blockquote></details>

Done.

---

*[Makefile, line 307 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-Xa--zAFPfLMMJ2UAb-r1-307:-L-Y4rXu0nCIIuvXy_Os:bon035c) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd4c98afbae85e29065be186779bbe81fb115b10/Makefile#L307)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Yeah, hadn't thought about the branch switching. Having a commit template is generally still really nice. It's a shame `git commit` fails spectacularly when the file isn't found.

As a compromise, perhaps adding to the warning in the `commit-msg`code could prove useful:

```bash
tplfile="${GIT_DIR}/.gitmessage"
if [ ! -e "${tplfile}" ]; then
  cat<<EOF
Run the following to install a sane commit message template for this repository:

    build/commit-template ${tplfile} && git config --local commit.template ${tplfile}
EOF   
fi
```
</blockquote></details>

Replaced with a commit hook.

---

*[build/commit-template, line 1 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-Y2tDtGnp-C3pFnsSA:-L-Y4wkxtsWx9nbZKjQP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd4c98afbae85e29065be186779bbe81fb115b10/build/commit-template#L1)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Might as well add the headline: `<package>: <short description>`
</blockquote></details>

Done.

---

*[build/commit-template, line 7 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-Y1lJPcDgnWDdvOvHZ:-L-YOyLPWI9AH0wcAnHY:baob2wc) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd4c98afbae85e29065be186779bbe81fb115b10/build/commit-template#L7)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Mind changing `CONTRIBUTING.md` to link to this file so that we don't duplicate?
</blockquote></details>

It's nice to have them denormalized so you can at least see the categories in CONTRIBUTING.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-:-L-Y4jX1F-oAZg7sngsF:buqy3i8)*

**Comment by tschottdorf** [comment]
:lgtm:, thanks for this!

> I'm planning to leave the lint out for a week or two to see how we do without it.

Do you think there's a scenario in which we don't need the lint? I don't think that's advisable since this only prints a warning and so the burden is on whoever compiles the commit notes to run the lint, which isn't a great process (compiling notes is annoying enough). I don't mind holding off for a week, though.

---

Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[build/commit-template, line 7 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-Y1lJPcDgnWDdvOvHZ:-L-YQz6bsNhvl0Ky6Ho7:b-m1oo0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd4c98afbae85e29065be186779bbe81fb115b10/build/commit-template#L7)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

It's nice to have them denormalized so you can at least see the categories in CONTRIBUTING.
</blockquote></details>

They'll rot, is more what I was going for, so even if you denormalize,  a link would be nice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-:-L-YR6CeGlA_leq2-B-T:b-539ahd)*

**Comment by benesch** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[githooks/prepare-commit-msg, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-YSWUoZXdHnhIRYxtE:-L-YSWUoZXdHnhIRYxtF:b4c0muw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0482d29d29a912477d428b6249e5367bdb1ffe8e/githooks/prepare-commit-msg#L9)):*
> ```
> fi
> 
> old=$(sed -n '1!p' "$1")
> ```

I think `tail -n +1` is a slightly more readable way of expressing "skip the first line", but definitely don't go through CI again for that.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-:-L-YSGqrjG8wNIL5lPVj:bnfp4nl)*

**Comment by benesch** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[githooks/prepare-commit-msg, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-YSWUoZXdHnhIRYxtE:-L-YSyQqEzuSJCq_Xn6_:bw4qioe) ([raw file](https://github.com/cockroachdb/cockroach/blob/0482d29d29a912477d428b6249e5367bdb1ffe8e/githooks/prepare-commit-msg#L9)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

I think `tail -n +1` is a slightly more readable way of expressing "skip the first line", but definitely don't go through CI again for that.
</blockquote></details>

`tail -n +2`, that is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447)*

**Comment by jordanlewis** [comment]
This might be controversial but I don't think we need the lint at all. Once people see the release note every time they make a commit, deleting it will take a conscious choice. That conscious choice may well be on purpose, in which case it's probably fine to have no release note for that commit.

I could be wrong about this, but that's why I wanted to wait a little - to see if we need the lint at all.

---

Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[githooks/prepare-commit-msg, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-YSWUoZXdHnhIRYxtE:-L-YTWZ0SRsyiB2nYPCb:bk1hi0y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0482d29d29a912477d428b6249e5367bdb1ffe8e/githooks/prepare-commit-msg#L9)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

`tail -n +2`, that is.
</blockquote></details>

I left this out because I don't think all `tail` implementations support that. I might be wrong? But I know that sed works for this everywhere.

---

*[build/commit-template, line 7 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-Y1lJPcDgnWDdvOvHZ:-L-YUMQSZZPZCQY0_LDK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd4c98afbae85e29065be186779bbe81fb115b10/build/commit-template#L7)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

They'll rot, is more what I was going for, so even if you denormalize,  a link would be nice.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-:-L-YTj0T-B7KYr28uQEq:b-9icjn4)*

**Comment by benesch** [comment]
Review status: 2 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[githooks/prepare-commit-msg, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-YSWUoZXdHnhIRYxtE:-L-YZQOFk2X2PBnK59Bv:b-b7tx8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0482d29d29a912477d428b6249e5367bdb1ffe8e/githooks/prepare-commit-msg#L9)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I left this out because I don't think all `tail` implementations support that. I might be wrong? But I know that sed works for this everywhere.
</blockquote></details>

(Support for it is actually mandated by POSIX: http://pubs.opengroup.org/onlinepubs/009695399/utilities/tail.html.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447)*

**Comment by benesch** [comment]
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[githooks/prepare-commit-msg, line 9 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447#-L-YSWUoZXdHnhIRYxtE:-L-YZh7tuyxGpRuED-9N:bkyrzie) ([raw file](https://github.com/cockroachdb/cockroach/blob/0482d29d29a912477d428b6249e5367bdb1ffe8e/githooks/prepare-commit-msg#L9)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

(Support for it is actually mandated by POSIX: http://pubs.opengroup.org/onlinepubs/009695399/utilities/tail.html.)
</blockquote></details>

(But again, don't go through CI for this!)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20447)*

**Comment by tschottdorf** [comment]
@jordanlewis then I don't see your point. If we never screw it up, what harm is in the lint? It's not in whoever composes the release notes' power to check whether everyone complies with the rules. It should be their prerogative to ensure that everyone is. And that is why we introduce a lint. It's not like folks screw these things up on purpose, it just happens.

Before you merge, can you double-check that

```
git commit -m 'storage: something

This is something that I do, and I may not be the only one.
It would be a shame if it broke.'
```

still works?

**Comment by jordanlewis** [comment]
Your example still works but it prints a warning because it doesn't have a release note.

**Comment by jordanlewis** [comment]
TFTR! I suppose the lint is fine then, I don't know. I wish you could specify a PR cover letter that would automatically get added to the merge commit.

**Pull request merged**